anvil_consortium_manager/anvil_audit.py

```diff
@@ -0,0 +1,158 @@
+"""Classes handling auditing of information in the Django database against AnVIL."""
+
+from abc import ABC, abstractproperty
+
+
+class AnVILAuditResults(ABC):
+    """Abstract base class to store audit results from AnVIL."""
+
+    @abstractproperty
+    def allowed_errors(self):
+        """List specifying the list of allowed errors for this audit result class."""
+        ...
+
+    def __init__(self):
+        self.verified = set()
+        self.errors = {}
+        self.not_in_app = set()
+
+    def add_not_in_app(self, record):
+        """Add a record that is on ANVIL but is not in the app.
+
+        Args:
+            record (str): An identifier for the record that is not in the app.
+                For example, for ManagedGroups, this will be the name of the group on AnVIL.
+        """
+        self.not_in_app.add(record)
+
+    def add_error(self, model_instance, error):
+        """Add an error for a Django model instance.
+
+        Args:
+            model_instance (obj): The Django model instance that had a detected error.
+            error (str): The error that was detected.
+
+        Raises:
+            ValueError: If the `error` is not in the `allowed_errors` attribute of the class.
+        """
+        if error not in self.allowed_errors:
+            raise ValueError("'{}' is not an allowed error.".format(error))
+        if model_instance in self.verified:
+            # Should this just remove it from verified and add the error instead?
+            raise ValueError(
+                "Cannot add error for model_instance {} that is already verified.".format(
+                    model_instance
+                )
+            )
```

Comment:
If trying to add an error for a model instance that is already in the `verified` set attribute of the class, should it raise a ValueError or silently remove it from the `verified` set? I can see arguments for both.

Reply:
Sticking with an exception for now.